tls/extensions/s2n_server_key_share.c

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "tls/extensions/s2n_server_key_share.h"
+
+#include "tls/s2n_client_extensions.h"
+#include "utils/s2n_safety.h"
+
+static int s2n_extensions_server_key_share_send_check(struct s2n_connection *conn);
+
+/*
+ * Check whether client has sent a corresponding curve and key_share
+ */
+static int s2n_extensions_server_key_share_send_check(struct s2n_connection *conn)
+{
+    const struct s2n_ecc_named_curve *server_curve, *client_curve;
+    server_curve = conn->secure.server_ecc_params.negotiated_curve;
```

Comment:
The  way group selection is done in this function requires to be updated with the changes suggested here https://github.com/awslabs/s2n/issues/1157. PS: I am working on a solution to this. 

However I do have a question on how we plan on supporting this case: 

- If the server has a group it prefers to the ones in the "key_share" extension but is still willing to accept the ClientHello,  it SHOULD send "supported_groups" to update the client's view of its preferences; this extension SHOULD contain all groups     the server supports, regardless of whether they are currently supported by the client.

Reply:
> If the server has a
   group it prefers to the ones in the "key_share" extension but is
   still willing to accept the ClientHello, it SHOULD send
   "supported_groups" to update the client's view of its preferences

This is for when supported_groups differs but the Server still accepts the ClientHello to reduce a roundtrip. In the case of #1157, we are only accepting connections with compatible supported_groups and do not have to exercise this case at the moment.

Reply:
I see, thanks for clarifying this use-case.